addons/mail/models/discuss/discuss_channel.py

```diff
@@ -1021,6 +1022,13 @@ def channel_pin(self, pinned=False):
         else:
             self.env['bus.bus']._sendone(self.env.user.partner_id, 'mail.record/insert', {"Thread": self._channel_info()[0]})
 
+    def _fetch_last_message(self, extra_domain=()):
+        return self.env['mail.message'].search(expression.AND([[
+            ("model", "=", "discuss.channel"),
+            ("res_id", "=", self.id),
+            ("message_type", "!=", "user_notification"),
+        ], extra_domain]), order="id DESC", limit=1)
```

Comment:
Why not use existing `_get_last_messages`?

Note: if it's only for `user_notification` I think we can ignore it. We never use that in channel, we should probably enforce it with a constraint + upgrade and get rid of it in code.
If we somehow want to support it, `_get_last_messages` should probably be fixed instead.

And `id < ...` can just be an extra param of it.

Reply:
Because `_get_last_messages` is not good from a SQL PoV.
The fact that it does a `MAX` means that it will need to still fetch and aggregate all the messages on the SQL server, even though the information is already in the index.

Reply:
So in a batch situation (which is the case for example in messaging menu) you suggest it's better to make 200 different queries rather than a single query using MAX?

Reply:
The best would be to do a lateral join, I can do that if you want

Reply:
Yes please, the goal would be to have a single method for last message (with optional param for id < value) that works in batch and that you consider efficient (benchmark with millions messages DB like our prod should be provided to confirm the change is indeed correct as it is very sensitive).

Reply:
Here is already the plan generated for both queries, using an random range of (sequential) channels.
Note that the queries are all run on a warmed up database. I would expect the difference in time to be even bigger on a cold database, especially for long channels.
The "Average" cases are taken on the last 10000 channels. The "Long Channel" case are taken on the channel with the most messages: 25982 messages.

GROUP BY
========

Average
-------

```sql
EXPLAIN (ANALYZE, VERBOSE)
SELECT res_id, MAX(mail_message.id) AS message_id
FROM mail_message                                                                                                       
WHERE model = 'discuss.channel' AND res_id IN (SELECT id FROM discuss_channel ORDER BY id DESC LIMIT 10000)
GROUP BY res_id;                                                                                                        
                                                                                           QUERY PLAN                                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=22581.61..23827.53 rows=124592 width=8) (actual time=121.958..124.186 rows=8937 loops=1)
   Output: mail_message.res_id, max(mail_message.id)
   Group Key: mail_message.res_id
   Batches: 1  Memory Usage: 7185kB
   ->  Nested Loop  (cost=610.70..21958.65 rows=124592 width=8) (actual time=6.795..111.012 rows=61577 loops=1)
         Output: mail_message.res_id, mail_message.id
         ->  HashAggregate  (cost=610.13..710.13 rows=10000 width=4) (actual time=6.764..9.076 rows=10000 loops=1)
               Output: discuss_channel.id
               Group Key: discuss_channel.id
               Batches: 1  Memory Usage: 913kB
               ->  Limit  (cost=0.42..485.13 rows=10000 width=4) (actual time=0.020..4.503 rows=10000 loops=1)
                     Output: discuss_channel.id
                     ->  Index Only Scan Backward using discuss_channel_pkey on public.discuss_channel  (cost=0.42..32068.34 rows=661594 width=4) (actual time=0.019..3.892 rows=10000 loops=1)
                           Output: discuss_channel.id
                           Heap Fetches: 5849
         ->  Index Only Scan using mail_message_model_res_id_id_idx on public.mail_message  (cost=0.57..2.00 rows=12 width=8) (actual time=0.005..0.010 rows=6 loops=10000)
               Output: mail_message.model, mail_message.res_id, mail_message.id
               Index Cond: ((mail_message.model = 'discuss.channel'::text) AND (mail_message.res_id = discuss_channel.id))
               Heap Fetches: 44146
 Planning Time: 0.422 ms
 Execution Time: 125.712 ms
(21 rows)
```

Note the following line:
```sql
 ->  Index Only Scan using mail_message_model_res_id_id_idx on public.mail_message  (cost=0.57..2.00 rows=12 width=8) (actual time=0.005..0.010 rows=6 loops=10000)
```
More specifically the `rows=6`, meaning that in average we fetched 6 rows per channel, where only one was enough. Even if it is an index only scan, it is still better to not fetch too many values from the index.

Long Channel
------------
```sql
EXPLAIN (ANALYZE, VERBOSE)
SELECT res_id, MAX(mail_message.id) AS message_id
FROM mail_message
WHERE model = 'discuss.channel' AND res_id = 706670
GROUP BY res_id;
                                                                              QUERY PLAN                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=0.57..2.99 rows=12 width=8) (actual time=6.189..6.190 rows=1 loops=1)
   Output: res_id, max(id)
   Group Key: mail_message.res_id
   ->  Index Only Scan using mail_message_model_res_id_id_idx on public.mail_message  (cost=0.57..2.81 rows=12 width=8) (actual time=0.037..4.073 rows=25982 loops=1)
         Output: model, res_id, id
         Index Cond: ((mail_message.model = 'discuss.channel'::text) AND (mail_message.res_id = 706670))
         Heap Fetches: 6
 Planning Time: 0.186 ms
 Execution Time: 6.217 ms
(9 rows)
```

LATERAL JOIN
============
Average
-------
```sql
EXPLAIN (ANALYZE, VERBOSE)                                                                                                                                                                                                          
           SELECT channel.id, latest_message.id                                                                                                                                                                                               
             FROM discuss_channel channel                                                                                                                                                                                                     
LEFT JOIN LATERAL (                                                                                                                                                                                                                           
                        SELECT id                                                                                                                                                                                                             
                          FROM mail_message                                                                                                                                                                                                   
                         WHERE model = 'discuss.channel'                                                                                                                                                                                      
                           AND res_id = channel.id                                                                                                                                                                                            
                      ORDER BY id DESC                                                                                                                                                                                                        
                         LIMIT 1                                                                                                                                                                                                              
                  ) latest_message ON TRUE                                                                                                                                                                                                    
            WHERE channel.id IN (SELECT id FROM discuss_channel ORDER BY id DESC LIMIT 10000);                                                                                                                                                
                                                                                           QUERY PLAN                                                                                                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------                                              
 Nested Loop Left Join  (cost=611.13..21051.08 rows=10000 width=8) (actual time=7.064..80.545 rows=10000 loops=1)                                                                                                                             
   Output: channel.id, mail_message.id
   ->  Nested Loop  (cost=610.56..13284.41 rows=10000 width=4) (actual time=7.037..28.667 rows=10000 loops=1)
         Output: channel.id
         Inner Unique: true
         ->  HashAggregate  (cost=610.13..710.13 rows=10000 width=4) (actual time=7.017..9.311 rows=10000 loops=1)
               Output: discuss_channel.id
               Group Key: discuss_channel.id
               Batches: 1  Memory Usage: 913kB
               ->  Limit  (cost=0.42..485.13 rows=10000 width=4) (actual time=0.022..4.759 rows=10000 loops=1)
                     Output: discuss_channel.id
                     ->  Index Only Scan Backward using discuss_channel_pkey on public.discuss_channel  (cost=0.42..32068.34 rows=661594 width=4) (actual time=0.021..4.201 rows=10000 loops=1)
                           Output: discuss_channel.id
                           Heap Fetches: 5849
         ->  Index Only Scan using discuss_channel_pkey on public.discuss_channel channel  (cost=0.42..1.27 rows=1 width=4) (actual time=0.002..0.002 rows=1 loops=10000)
               Output: channel.id
               Index Cond: (channel.id = discuss_channel.id)
               Heap Fetches: 5849
   ->  Limit  (cost=0.57..0.76 rows=1 width=4) (actual time=0.005..0.005 rows=1 loops=10000)
         Output: mail_message.id
         ->  Index Only Scan Backward using mail_message_model_res_id_id_idx on public.mail_message  (cost=0.57..2.81 rows=12 width=4) (actual time=0.005..0.005 rows=1 loops=10000)
               Output: mail_message.id
               Index Cond: ((mail_message.model = 'discuss.channel'::text) AND (mail_message.res_id = channel.id))
               Heap Fetches: 6446
 Planning Time: 0.458 ms
 Execution Time: 81.101 ms
(26 rows)
```

In comparison with the previous, we have here
```sql
->  Index Only Scan Backward using mail_message_model_res_id_id_idx on public.mail_message  (cost=0.57..2.81 rows=12 width=4) (actual time=0.005..0.005 rows=1 loops=10000)
```
With `rows=1` because we got exactly 1 or 0 messages per channel thanks to the limit.
The number of channels I too here is 10000 but the difference is the same no matter how many channels we take into account.


Long Channel
------------
```sql
EXPLAIN (ANALYZE, VERBOSE)
           SELECT channel.id, latest_message.id
             FROM discuss_channel channel
LEFT JOIN LATERAL (
                        SELECT id
                          FROM mail_message
                         WHERE model = 'discuss.channel'
                           AND res_id = channel.id
                      ORDER BY id DESC
                         LIMIT 1
                  ) latest_message ON TRUE
            WHERE channel.id = 706670;
                                                                                   QUERY PLAN                                                                                    
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop Left Join  (cost=1.00..3.22 rows=1 width=8) (actual time=0.068..0.070 rows=1 loops=1)
   Output: channel.id, mail_message.id
   ->  Index Only Scan using discuss_channel_pkey on public.discuss_channel channel  (cost=0.42..2.44 rows=1 width=4) (actual time=0.030..0.031 rows=1 loops=1)
         Output: channel.id
         Index Cond: (channel.id = 706670)
         Heap Fetches: 0
   ->  Limit  (cost=0.57..0.76 rows=1 width=4) (actual time=0.035..0.035 rows=1 loops=1)
         Output: mail_message.id
         ->  Index Only Scan Backward using mail_message_model_res_id_id_idx on public.mail_message  (cost=0.57..2.81 rows=12 width=4) (actual time=0.033..0.033 rows=1 loops=1)
               Output: mail_message.id
               Index Cond: ((mail_message.model = 'discuss.channel'::text) AND (mail_message.res_id = channel.id))
               Heap Fetches: 0
 Planning Time: 0.323 ms
 Execution Time: 0.105 ms
(14 rows)
```

Reply:
Thank you for the details!

Will you push this version to the PR?
Maybe it should be done separately than fixing the serialization?